### PR TITLE
[Draft] RFC: Generalized Type Ascription

### PR DESCRIPTION
**Pull request by Centril**
This is a draft version of an RFC for you to review, before a formal proposal is made for consideration.

# Summary
[summary]: #summary

[RFC 803]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md

This RFC supersedes and subsumes [RFC 803]. We finalize a general notion of type ascription uniformly in patterns, expressions, `let` bindings, and `fn` definitions. You may now for example write:

```rust
let x = (0..10).collect() : Vec<_>;

do_stuff(try : Option<u8> { .. });

do_stuff(async : u8 { .. });

do_stuff(unsafe : u8 { .. });

do_stuff(loop : u8 { .. });

let alpha: u8 = expr;
    ^^^^^^^^^

let [x: u8, y, z] = stuff();
    ^^^^^^^^^^^^^

if let Some(beta: u8) = expr { .. }
            ^^^^^^^^

fn foo(Wrapping(alpha: usize)) {}
       ^^^^^^^^^^^^^^^^^^^^^^
```

Here, the underlined bits are patterns. Note however that this RFC does *not* introduce global type inference.

Finally, we lint (warn-by-default) when a user writes `Foo { $field: $ident : $type }` and the compiler instead suggests: `Foo { $field: ($ident : $type) }`.

### CODE REVIEW

File: text/0000-generalized-type-ascription.md [hunk 1]
```diff
@@ -0,0 +1,1563 @@
...
+   While this may change in the future, it may also be the case that anonymous
+   `arg: impl Trait`s can't ever be turbofished. In such a case, type ascription
+   is our saving grace; it works independently of how `display_all` is defined
+   and let's us easily constrain the type of `elems` in a syntactically
+   light-weight way in any case.
+
+6. Type ascription is helpful when doing *type* driven development (TDD)
+   and opens up more possibilities to move in the direction of
+   interactive development as is possible with [agda-mode] and [idris-mode].
+
```
Comment (nrc) [inline text/0000-generalized-type-ascription.md]: ANother benefit (which I don't think was ni 803) is that when you get a gnarly type error, type ascription can quickly help you understand and narrow down the error.

File: text/0000-generalized-type-ascription.md [hunk 2]
```diff
@@ -0,0 +1,1563 @@
...
+   fn display_all<I: Iterator<Item: Display>>(elems: I) { ... }
+   ```
+
+   While this may change in the future, it may also be the case that anonymous
+   `arg: impl Trait`s can't ever be turbofished. In such a case, type ascription
+   is our saving grace; it works independently of how `display_all` is defined
+   and let's us easily constrain the type of `elems` in a syntactically
+   light-weight way in any case.
+
+6. Type ascription is helpful when doing *type* driven development (TDD)
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md]: nit: When I google "TDD" I get _test_ driven development, as I expected, so I would discourage this abbreviation.

File: text/0000-generalized-type-ascription.md [hunk 3]
```diff
@@ -0,0 +1,1563 @@
...
+}
+```
+
+In the last case, the typing annotation is both *most local* and also does not
+require you to annotate information that is both obvious to the reader
+(who is familiar with `Option<T>`) and to the compiler
+(that `expr : Option<?T>` for some `?T`).
+Because the annotation is more local, we can employ more local reasoning.
+This is particularly useful if the `enum` contains many variants in which
+case the type ascription on `expr` may not be immediately visible.
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md:179]: I really like this example, and would like to see the multi-variant thing more.  Maybe something like

```rust
match foo.parse() {
    Ok(x: i32) => ...,
    Err(e) => ...
}
```

as opposed to `::<i32>` or `: Result<i32, _>`.

File: text/0000-generalized-type-ascription.md [hunk 4]
```diff
@@ -0,0 +1,1563 @@
...
+  |
+  |                alpha: usize
+  |                beta: bool
+  |
+
+error: aborting due to previous error
+```
+
+This also gives the compiler an opportunity to tell you what the types are
+if it so happens that you need this help.
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md]: I think this would be hard right now unless it's a trait impl, so you might want to have the code example explicitly be such, or change the error to just say that you need types, without suggesting them.  (Especially since diagnostic quality is something that doesn't need RFCing -- people will just make it better even if the RFC doesn't say so, in the cases that are feasible.)

File: text/0000-generalized-type-ascription.md [hunk 5]
```diff
@@ -0,0 +1,1563 @@
...
+// Unconstrained type variable ?T of Wrapping<?T>:
+fn foo(Wrapping(value)) -> usize { ... }
+```
+
+By using this mechanism, we gain a measure of elision and can make writing
+more ergonomic.
+
+[RFC 1685]: https://github.com/rust-lang/rfcs/blob/master/text/1685-deprecate-anonymous-parameters.md
+
+#### [RFC 1685] and deprecation schedule
```
Comment (nrc) [inline text/0000-generalized-type-ascription.md:816]: Could you say what RFC 1685 is so I don't need to click through?

File: text/0000-generalized-type-ascription.md [hunk 6]
```diff
@@ -0,0 +1,1563 @@
...
+```
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There are none.
+
+# Possible future work
+[possible future work]: #possible-future-work
+
```
Comment (nrc) [inline text/0000-generalized-type-ascription.md]: Not sure whether you want to think about it now, or leave it for the future, but I'd love to be able to use `impl Trait` in type ascription. I'm not sure what the semantics would be though (I'd like `e: impl T` to mean that the type of `e` implements `T`, but I'm not sure how that would interact with existing uses in function args and return types).

**Review by nrc — COMMENTED** [review]
Big :+1: from me!

It's probably worth discussing the implementation status of #803 in more detail, iirc, non-coercing ascription is implemented, but coercions are not. (As an alternative design, it might be worth considering different ways that coercion and ascription interact. Some people have wanted to keep ascription meaning no coercion, and only coerce with `as`, however, this doesn't work for some coercions from fixed size to dynamically sized types).

File: text/0000-generalized-type-ascription.md [hunk 7]
```diff
@@ -0,0 +1,1563 @@
...
+}
+```
+
+and against:
+
+```rust
+match expr {
+    None => logic,
+    Some(vec: Vec<u8>) => logic,
+}
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md:170]: If we mix this with match ergonomics, the ascribed type of `vec` should be `&Vec<u8>` (or `&[u8]` since coercion is allowed)?

```rust
match &expr {
    None => logic,
    Some(vec: &Vec<u8>) => logic,
}
```

File: text/0000-generalized-type-ascription.md [hunk 8]
```diff
@@ -0,0 +1,1563 @@
...
+interpreted as `(foo : bar).quux()` and `(foo : bar).quux`. 
+
+However, this does not mean that the operator `:` should bind more tightly than
+operators such as the unary operators `-`, `*`, `!`, `&`, and `&mut`.
+In particular, for the latter two operators, we expect that if someone writes
+`&x : Type`, it would be interpreted as `(&x) : Type` as opposed to `&(x : Type)`.
+
+Instead, we propose that whenever type ascription is followed by a
+field projection or a method call, the projections or the call should apply
+to the entire ascribed expression.
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md:559]: Perhaps clarify that:

```rust
&x.y:z    => (&(x.y)):z
&a:b.c    => &((a:b).c)
```

File: text/0000-generalized-type-ascription.md [hunk 9]
```diff
@@ -0,0 +1,1563 @@
...
+You may type ascribe these blocks in the same way as above:
+
+```rust
+let future = async : io::Result<()> {
+    ...
+};
+```
+
+Do note however that in this case, you are annotating the inner type of the
+resulting future and not the future itself. Thus, this is equivalent to:
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md:625]: I think having this inconsistency here is not a good idea. I'd rather choose a different symbol just for `async` (`async -> io::Result<()>`) or just don't add this feature to `async` block.

File: text/0000-generalized-type-ascription.md [hunk 10]
```diff
@@ -0,0 +1,1563 @@
...
+
+if let Thing { field: binding: MyType } = make_thing() {
+    ...
+}
+```
+
+### In macros
+
+Just as we noted before that type ascription work in expression macros so may
+you use type ascription in pattern macros. For example:
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md:667]: You may also note that `:` is not in the follow set of `pat`, meaning code like

```rust
macro_rules! a {
    ($p:pat: u32) => {}
}
```

will not compile already 

```
error: `$p:pat` is followed by `:`, which is not allowed for `pat` fragments
 --> src/main.rs:2:12
  |
2 |     ($p:pat: u32) => {}
  |            ^
```

thus allowing us to extend `pat` to accept ascription without breaking backward compatibility.

File: text/0000-generalized-type-ascription.md [hunk 11]
```diff
@@ -0,0 +1,1563 @@
...
+
+let ref b: X = a; // Note the type!
+let &c : &X = b; // And here!
+```
+
+This entails for example that a Rust compiler will interpret `ref x : T` as
+`ref (x : T)` instead of `(ref x) : T`. The same applies to `ref mut`, `&`,
+and `&mut`.
+
+The grammar of `let` bindings is changed from:
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md]: This should also apply to `if let` and `while let` (assuming they are still three distinct rules).

File: text/0000-generalized-type-ascription.md [hunk 12]
```diff
@@ -0,0 +1,1563 @@
...
+// The type of `g3: for<'a> fn(Wrapping<&'a usize>) -> ()`.
+fn g2(Wrapping(x: &usize)) {}
+
+// Same here. Determined by `x: T`.
+// The type of `g2: for<T> fn(Wrapping<T>) -> ()`.
+fn g3<T>(Wrapping(x: T)) {}
+
+// A type variable is induced by `impl Display`
+// and then `typeof(x)` is that variable.
+// The type of `g3: for<T: Display> fn(Wrapping<T>) -> ()`.
```
Comment (kennytm) [inline text/0000-generalized-type-ascription.md]: The function names in the comments are all incorrect :)

File: text/0000-generalized-type-ascription.md (hunk 7 above)
Comment (scottmcm) [reply text/0000-generalized-type-ascription.md:170]: That makes me think of another motivation for this: it helps address some of the concerns raised in https://internals.rust-lang.org/t/lived-experiences-strange-match-ergonomics/7817/17?u=scottmcm

Want to know whether `let x =` is a reference?  Type ascribe it in the pattern.
What to know whether `let Struct { x } =` is a reference?  Type ascribe it in the pattern.

Edit: posted to https://internals.rust-lang.org/t/lived-experiences-strange-match-ergonomics/7817/35?u=scottmcm

**Comment by kennytm** [comment]
The issue currently blocking type ascription (rust-lang/rust#23416) is soundness issue when coercion happens:

> Ensure soundness with respect to `ref` positions: 
>
> - [ ] `let ref x = <expr>` / `let mut ref = <expr>`
> - [ ] `match <expr>: Type { ref x => ... }`
> - [ ] `(<expr>: Type).method_with_ref_self()`

The issue is elaborated in https://github.com/rust-lang/rfcs/pull/803#issuecomment-78102035 and https://github.com/rust-lang/rust/issues/23416#issuecomment-315785799. It seems this RFC did not address this issue (can't find the word "sound" and this RFC does support coercion).

File: text/0000-generalized-type-ascription.md [hunk 13]
```diff
@@ -0,0 +1,1563 @@
...
+
+Furthermore, it was also noted on the same PR by @nikomatsakis that:
+
+> [..] we don't think there's much use of this in the wild,
+> and we have an automated fix.
+
+Thus, we believe it is reasonable to expedite the schedule.
+
+# Prior art
+[prior-art]: #prior-art
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md]: SML also uses exactly the `expr: type` syntax (https://courses.cs.washington.edu/courses/cse341/04wi/lectures/02-ml-intro.html) and even the `pat ::= pat : ty` grammar production (http://sml-family.org/sml97-defn.pdf page 9).

File: text/0000-generalized-type-ascription.md (hunk 3 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:179]: This is a great example; I'll be sure to include it.

File: text/0000-generalized-type-ascription.md [hunk 14]
```diff
@@ -0,0 +1,1563 @@
...
+struct X {}
+let a = X {};
+
+let ref b: X = a; // Note the type!
+let &c : &X = b; // And here!
+```
+
+This entails for example that a Rust compiler will interpret `ref x : T` as
+`ref (x : T)` instead of `(ref x) : T`. The same applies to `ref mut`, `&`,
+and `&mut`.
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md]: Are you certain it's the same between `ref` and `&`?  In

```rust
let &x : &i32 = &0;
```

the `x` is an `i32`, so I think the parse is `let (&x) : &i32`, not `let &(x: &i32)`.

File: text/0000-generalized-type-ascription.md (hunk 3 above)
Comment (scottmcm) [reply text/0000-generalized-type-ascription.md:179]: Update looks great.  Thanks!

File: text/0000-generalized-type-ascription.md (hunk 12 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md]: Thanks for spotting this! Fixed :)

File: text/0000-generalized-type-ascription.md (hunk 5 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:816]: Sure thing; I hope the added bit is satisfactory :)

File: text/0000-generalized-type-ascription.md (hunk 6 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md]: Good catch; I think I can add it now.
I think the semantics should be rather well defined:

1. In `fn foo(Wrapping(x: impl Trait)) { .. }` it acts just like `universal_impl_trait` does otherwise.

2. In the function body it acts like `let x: impl Trait = ..;` does. The semantics of this is defined by https://github.com/rust-lang/rfcs/pull/2071.

File: text/0000-generalized-type-ascription.md (hunk 13 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md]: Excellent! I looked at OCaml and was disheartened that it didn't have this so I didn't bother to check SML. But this is great :)

File: text/0000-generalized-type-ascription.md (hunk 8 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:559]: Good idea; added a paragraph :)

File: text/0000-generalized-type-ascription.md (hunk 10 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:667]: Good idea; added a note :)

File: text/0000-generalized-type-ascription.md (hunk 11 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md]: It automatically applies to `if let` and `while let` because they already used the `pat` grammar, see https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L1761-L1764 (could be outdated but I don't think it is in this case).

File: text/0000-generalized-type-ascription.md (hunk 4 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md]: Nuanced the language a bit here to make it optional :)

File: text/0000-generalized-type-ascription.md (hunk 7 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:170]: This is great; I added a section in the motivation about it :)

File: text/0000-generalized-type-ascription.md (hunk 9 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:625]: I'll keep it for now but add an unresolved question about it. :)

At the moment I think that while it is unfortunate that we can't be fully consistent
about the `KEYWORD : Type { ... }` => `KEYWORD { ... } : Type` desugaring,
this problem is really inherent to the nature of `async fn` using the inner-type method itself.
Thus, I think that `async : R { .. }` desugaring as `async { .. } : impl Future<Output = R>` aligns with what people will expect.

Comment (kennytm) [reply text/0000-generalized-type-ascription.md:625]: ((or we could make `async fn` return the outer type 🙃))

File: text/0000-generalized-type-ascription.md [hunk 15]
```diff
@@ -0,0 +1,1740 @@
...
+   previously needed to. This especially holds in pattern contexts.
+   This will be made clear later on in this RFC.
+
+5. Turbofish is not always possible! Consider for example:
+
+   ```rust
+   fn display_all(elems: impl Iterator<Item: Display>) { ... }
+   ```
+
+   As of today (2018-08-06), it is not possible to use turbofish at all
```
Comment (varkor) [inline text/0000-generalized-type-ascription.md:99]: Still true as of 2018-08-07 😁

File: text/0000-generalized-type-ascription.md [hunk 16]
```diff
@@ -0,0 +1,1740 @@
...
+```rust
+let x = (0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>
+    .unwrap()
+    .map(other_computation) : Vec<usize>
+    .into() : Rc<[_]>;
+```
+
+To that end, `foo : bar.quux()` and `foo : bar.quux` should unambiguously be
```
Comment (varkor) [inline text/0000-generalized-type-ascription.md:549]: Note that this is unintuitive if Rust ever has type-level functions, as there one would naturally expect `bar.quux()` to be associated the most tightly.

File: text/0000-generalized-type-ascription.md [hunk 17]
```diff
@@ -0,0 +1,1740 @@
...
+```rust
+let x = (((0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>)
+    .unwrap()
+    .map(other_computation) : Vec<usize>)
+    .into() : Rc<[_]>;
+```
+
+We suggest instead that you should be able to write:
```
Comment (varkor) [inline text/0000-generalized-type-ascription.md:538]: This is readable if line-separated, or with extra parentheses, but this assumes the user is doing that (or running rustfmt regularly). If not, then the precedence suggested here is confusing.

File: text/0000-generalized-type-ascription.md [hunk 18]
```diff
@@ -0,0 +1,1740 @@
...
+of the compiler.
+
+However, in some cases, you can determine the type from the pattern alone.
+Therefore, when the type is fully determined you may omit the type ascription.
+Thus, the following definitions are legal:
+
+```rust
+fn foo<T>(Wrapping(value: T)) -> usize { ... }
+
+fn bar(Wrapping(value: usize)) -> usize { ... }
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md:789]: While I really like this feature, it might be worth scoping it down initially to avoid needing to specify all the details.  For example, there are no examples of lifetimes, thus far.  (And I think it'd work fine, but there are a bunch of questions around things like lifetime elision if you have `Foo { x: x: &i32, y: y: &i32 }` and it's `Foo<'_>`.)

Maybe start with just the "no generics" case?  (Like `Quux` below.)

File: text/0000-generalized-type-ascription.md [hunk 19]
```diff
@@ -0,0 +1,1740 @@
...
+[RFC_803_motivation]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md#motivation
+
+[pointfree persuasion]: https://en.wikipedia.org/wiki/Tacit_programming
+
+[TwoHardThings]: https://martinfowler.com/bliki/TwoHardThings.html
+
+Type ascription is useful. A motivation for the feature is noted in the merged,
+but thus far not stabilized, [RFC 803][RFC_803_motivation] which introduces
+type ascription in expression contexts as `expr : T`. We reinforce that RFC
+with more motivation:
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md:60]: These are all good motivations, but I think they vary in strength.  As a nit, consider reordering the strongest ones first.  Obviously that's subjective, but right now (1) feels like it'll get "one variable just isn't that big of a deal" push-back.  I really like (4), for example, and (2) is IMHO a stronger phrasing of the annoyance.

File: text/0000-generalized-type-ascription.md (hunk 16 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:549]: Thanks; added a section in the rationale discussing this.

File: text/0000-generalized-type-ascription.md (hunk 17 above)
Comment (Centril) [reply text/0000-generalized-type-ascription.md:538]: And also this in the same section :)

File: text/0000-generalized-type-ascription.md [hunk 20]
```diff
@@ -1526,6 +1526,55 @@ Furthermore, it was also noted on the same PR by @nikomatsakis that:
...
+    .map(some_computation)
+    .collect() : Foo
+    .unwrap()
+    .map(other_computation) : Bar
+    .into() : Baz;
+```
+
+However, if you write this on a single line, or simply consider `x : T.foo()`
+a user might parse this as `x : (T.foo())` instead.
+While Rust does not have any "type level methods", and wherefore this parse
```
Comment (varkor) [inline text/0000-generalized-type-ascription.md]: "wherefore" isn't quite right here ("therefore" is closer, but still awkward). Maybe rephrase to something like: "While at this stage Rust does not support type-level methods (meaning this parse currently makes no sense), a user may nonetheless make this mistake."

File: text/0000-generalized-type-ascription.md (hunk 18 above)
Comment (scottmcm) [reply text/0000-generalized-type-ascription.md:789]: ...or you could just define the full unification algorithm :rofl:

File: text/0000-generalized-type-ascription.md [hunk 21]
```diff
@@ -0,0 +1,1956 @@
...
+//    b) checks that `x : &'a i32`.
+//    c) this entails that `'?b = 'a` and so `'?b` is substituted for `'a`.
+//
+// 3) Looking at `$pat_2 = x: Bar<'_>`, the compiler:
+//    a) infers that `x : Bar<'?c>` and adds unification variable `'?c`.
+//    b) checks that `x : Bar<'a>`.
+//    c) this entails that `'?c = 'a` and so `'?c` is substituted for `'a`.
+//
+// 4) Thus, `good_11 : for<'a> fn(Wibble<'a>) -> ()`.
+fn good_11(Wibble(x: &i32, y: Bar<'_>)) {}
```
Comment (scottmcm) [inline text/0000-generalized-type-ascription.md]: Wow, this section is awesome!

I think I agree with all of them except for `good_11`: I would expect that 
```rust
fn good_11(Wibble(x: &i32, y: Bar<'_>)) {}
```
is equivalent to
```rust
fn good_11(Wibble(x: &'_ i32, y: Bar<'_>)) {}
```

is equivalent to 
```rust
fn good_11<'a, 'b>(Wibble(x: &'a i32, y: Bar<'b>)) {}
```

Which is illegal since it, unlike `Wibble`, doesn't require that `'a == 'b`.

That's because the version not wrapped in `Wibble`:
```
fn rutabaga(x: &i32, y: Bar<'_>) {}
```

would have two lifetimes, so I think the wrapped version should too.

This helps local reasoning, as you don't need to look at the declaration of Wibble to know whether the two lifetimes are the same or different (since `Product` and `Wibble` cases look the same in the function header).

--

So maybe then there's 

```rust
fn good_11<'a>(Wibble(x: &'a i32, y: Bar<'a>)) {}
```

and

```rust
fn bad_4(Wibble(x: &i32, y: Bar<'_>)) {} //~ ERROR insufficiently constrained
```

Comment (Centril) [reply text/0000-generalized-type-ascription.md]: Glad you think so :)

I agree that `good_11` is the tricky one. The rule you imposed certainly does improve local reasoning and the expectation induced by `rutabaga` does seem like it would happen to folks sometimes.

However, I do wonder if people care and think so much about lifetime parameters (if you are dealing with unsafe code where this is important you should quantify explicitly) that it matters in practice. I mostly don't think about them that much personally.

+ At the call site `good_11` won't be callable with something where given `&'a i32` and `Bar<'b>` you have that `'a != 'b` because you won't be able to construct such a `Wibble<'?>` in the first place.

+ What remains is then confusion at the definition site of `good_11` where the user may think that `'a != 'b`. However, when the user actually tries to make use of this mistaken belief, the compiler should be able to give the user the actual signature of the function and/or say that the lifetimes are really the same. Thus, the user can therefore "learn on the job".

In any case, it is also true that your proposed rule allows strictly less than my rule so if we use your rule and then change our minds we can switch. So we are forward compatible.

Finally and as an aside, there are two ways to make your changed `good_11` shorter:

```rust
// in-band lifetimes (assuming we don't ditch it)
fn good_11(Wibble(x: &'a i32, y: Bar<'a>)) {}

// The types are fully determined by `Wibble<'_>` itself.
fn good_11(Wibble(x, y)) {}
```

EDIT: Perhaps the last one is so ergonomic that `fn good_11(Wibble(x: &i32, y: Bar<'_>))` simply doesn't occur for it to matter?

I'll go and add an unresolved question for now; I think this is something which is subtle and where we'll want to weigh the trade-offs and discuss them on the RFC. :)

Comment (scottmcm) [reply text/0000-generalized-type-ascription.md]: I think that
```rust
fn good_11(Wibble(x, y)) {}
```
should be rejected for the same reason that we now require `-> Foo<'_>` instead of just `-> Foo`: the existence of a lifetime should be visible in the function header.

So I still think the answer is that elision fills in lifetimes before the unification algorithm runs.  and then the `fn arugula(Wibble(x, y)) {}` example an error because the unification algorithm doesn't find anything for the lifetime variables to unify with, and thus the signature is under-constrained.

Comment (scottmcm) [reply text/0000-generalized-type-ascription.md]: ...and I thought of more stuff for these examples: match ergonomics.

I think the following would work?
```rust
fn good_13(Foo(x): &_) {}
```

Giving a type of `for<'a> fn(&'a Foo)` and, in the body, `x: &'a usize`.

File: text/0000-generalized-type-ascription.md [hunk 22]
```diff
@@ -1715,6 +1900,271 @@ this method variant could only exist in the core library.
...
+We argue that consistency of function definitions with closures
+is paramount for an understandable language.
+Thus, we conclude that match ergonomics should apply.
+
+### How much unification do we permit?
+
+One of the rules we've imposed is that each lifetime position mentioned
+in a type fragment of some pattern in a function parameter introduces
+a distinct lifetime. However, this means that type inference may sometimes
+introduce two many lifetimes for one type and therefore reject the definition.
```
Comment (varkor) [inline text/0000-generalized-type-ascription.md]: *too many

**Comment by Centril** [comment]
Thanks all for chipping in! Time to publish :)
